### DISCUSSION TITLE
[use case] Get column average to command line

### DISCUSSION THREAD

**Discussion by abitrolly** [saulpw/visidata#1516] [General]
`visidata` is awesome. Want to make it even better. )

Given this JSON - https://gitlab.com/gitlab-org/gitlab-docs/-/pipelines.json - I need to calculate column average for build duration.

I can extract the column using `jq` (and filter it, but I'd like to keep it simple).
```bash
$ curl -sSL https://gitlab.com/gitlab-org/gitlab-docs/-/pipelines.json | jq ".pipelines[].details.duration"            
null
1022
1010
1064
1071
1055
924
1039
960
962
1058
1065
1048
980
1014
```
And then use `awk` or Python to process it, but I would prefer to use `visidata` from command line as it is more powerful in a long run.

So the sequence of keys to get average is`#` `z+` `avg` `<enter>`. And then there is a line `_avg=1019.43` at the bottom of the window, which is impossible to copy/paste anywhere.

The most straightforward approach I can think of is to add these as macros.

```
curl ... | vd -m '# z + a v g ENTER' --batch`
```

And yes, then there is no `-m, --macro` command, and it is not evident here that the result should be written to `stdout`.

**Comment by saulpw** [comment]
Hi @abitrolly, this is an interesting idea and I've been thinking about how to make it easier to do things like this.  I'll give your `-m` option some consideration.  Maybe more like `vd foo.jsonl -m $'#z+avg\n'`.

As a workaround for right now, you can save a cmdlog with those commands, and then run it in batch mode, where every status msg is sent to stderr:

    vd -b -p pipeline-duration-avg.vdj 2>&1 | grep avg

**Reply by abitrolly** [reply]
> Maybe more like `vd foo.jsonl -m $'#z+avg\n'`.

What does `$` sign do here?
I see the problem to enter shortcuts like Alt+Ctrl+M. In my case it will be 'AltCtrlM`.

It is also interesting if it is possible to produce compiling grammar for macros, to see which keys are defined and which are possible from which state. Maybe all states where string is entered don't react to special keys.

How are macros stored? Is it already a string like the one above? I can not validate because of https://github.com/saulpw/visidata/issues/1513

It would be nice to have some converter between macros and readable commands (function calls) to see if it is possible to have a command batch syntax that could be written by hand. Which might become yet another `awk` scripting language. :D

> 
> As a workaround for right now, you can save a cmdlog with those commands, and then run it in batch mode, where every status msg is sent to stderr:
> ```
> vd -b -p pipeline-duration-avg.vdj 2>&1 | grep avg
> ```

Just for the record. `Shift+D` to review `cmdlog` and `Ctrl+D` to save it.

```
✗ cat ppl2.json | ~/bin/vd -b -p ppl2_cmdlog.vdj 2>&1 | grep avg
text_avg=1019.43
```
Not everybody knows what a `stderr` redirection is. ) Why not to print `avg=1019.43` to `stdout`?

**Reply by saulpw** [reply]
> What does $ sign do here?

It's a shell-ism to properly convert escape sequences (like \n into newline).

> How are macros stored?

As little cmdlogs.

> It would be nice to have some converter between macros and readable commands (function calls) to see if it is possible to have a command batch syntax that could be written by hand.

Macros are just commands.  They are as readable as cmdlogs already are.  I've been thinking about making a simple cmdlog so that it could become this 'command language' but it wouldn't be anything more than just "longname input" on each line.

> Why not to print avg=1019.43 to stdout?

Because that's where the final sheet is output if you are using vd in a pipeline.

**Reply by abitrolly** [reply]
> As little cmdlogs.

I need a reference to see how `little cmdlog` format is different from just `cmdlog`. Google doesn't find anything https://www.google.com/search?q=visidata+cmdlog+format

The GMIC program has a lot of commands and it exposes an interface like this https://gmic.eu/tutorial/basics.html

> Because that's where the final sheet is output if you are using vd in a pipeline.

Maybe some flag `--status-only` or `--json` to output something that can be filtered more easily. The problem with `avg=1019.43` is that it is not easy to leave only `1019.43`.
